source/plugin/migrating-from-7-to-8.rst

```diff
@@ -0,0 +1,221 @@
+    org.spongepowered.api.world.Location
+    org.spongepowered.api.world.World
+    org.spongepowered.api.world.client.ClientLocation
+    org.spongepowered.api.world.client.ClientWorld
+    org.spongepowered.api.world.server.ServerLocation
+    org.spongepowered.api.world.server.ServerWorld
+    org.spongepowered.plugin.builtin.jvm.Plugin
+
+SpongeAPI 8 is a significally upgraded API compared to SpongeAPI 7, such that every plugin will need to be updated to
+be compatible. While we cannot list every little thing that you will need to change here, this page contains some of
+the more common migrations that will be required.
+
+.. note::
+    `We provide a plugin template that you can clone to create your own plugins <https://github.com/SpongePowered/sponge-plugin-template>`__.
+    While this guide is primarily intend for migrating plugins, you may find it useful to investigate this template to
+    help your migration, particularly with the plugin metadata changes.
+
+``@Plugin`` annotation and migration from ``mcmod.info`` to ``sponge_plugins.json``
+===================================================================================
+
+SpongeAPI 8 introduces a new :javadoc:`Plugin` annotation that only contains the plugin ID. As a result, the SpongeAPI
+build no longer contains an annotation processor that will generate the metadata file for you.
+
+To generate the metadata file, you can either:
+
+- Create the file yourself by creating ``sponge_plugins.json`` in your resources root and filling it out with the required
+  information
+- Use SpongeGradle 2 and define the metadata in the buildscript `as in this example 
+  <https://github.com/SpongePowered/sponge-plugin-template/blob/88d3c35853a687a7dc1540db43a9f9a135c03819/build.gradle.kts#L16-L40>`__
+
+More information about the metadata file can be found at :doc:`plugin-meta`.
+
+Game Lifecycle
+==============
+
+The previous lifecycle events (such as ``GamePreInitializationEvent``) have been removed. SpongeAPI 8 uses a clearer 
+lifecycle so you know exactly when to register various parts of your plugin. Most of your plugin's setup should now be
+done in "Register" events, such as :javadoc:`RegisterCommandEvent` or :javadoc:`RegisterDataEvent`, though there are a
+few generic events (such as :javadoc:`LoadedGameEvent`) that may also be used.
+
+More information about the game lifecycle can be found at :doc:`lifecycle`.
+
+Engines
+=======
+
+SpongeAPI 8 introduces the concept of an engine. While SpongeAPI 7 was mostly designed for servers, SpongeAPI 8 considers
+the client as a first class citizen. The :javadoc:`Server` and :javadoc:`Client` are both :javadoc:`Engine`.
+
+In general plugin development, it is likely that you will only really consider the server - this is true even in 
+singleplayer environments as the game client starts a singleplayer server. However, be aware that there are times when
+running on the client where the server engine does not exist.
```

Comment:
There's something off about the last sentence, perhaps "There can be times when the client is running but no server engine exists" instead?